### PR TITLE
Creating Container structure test task

### PR DESCRIPTION
**Pull request by navin22**
- Runs container structure tests. Uses the framework and runner provided [here](https://github.com/GoogleContainerTools/container-structure-test).
- As of now downloading every time on the fly. will try to optimize it later.
- Taking only single file as input.

### CODE REVIEW

File: Tasks/ContainerStructureTestV1/task.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    "version": {
+        "Major": 1,
+        "Minor": 0,
+        "Patch": 0
+    },
+    "demands": [],
+    "minimumAgentVersion": "2.0.0",
+    "inputs": [
+        {
+            "name": "testFile",
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV1/task.json]: Change this to configFile.

File: Tasks/ContainerStructureTestV1/task.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            "type": "multiline",
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV1/task.json]: Should be filePath type input. Refer to KubernetesManifest task. Even if we want to add support for Glob pattern tommorow, it should be possible

File: Tasks/ContainerStructureTestV1/task.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            "label": "Test Files",
+            "required": true,
+            "helpMarkDown": "Test files path, that contains container structure tests. Either .yaml or .json files"
+        },
+        {
+            "name": "image",
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV1/task.json]: Check Docker task for inputs - Use 3 inputs - dockerRegistryServiceConnection (in Docker task it's called containerRegistry; But call the input dockerRegistryServiceConnection as we got some feedback), repository and tags.

File: Tasks/ContainerStructureTestV1/task.loc.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  "helpUrl": "https://docs.microsoft.com/azure/devops/pipelines/tasks/test/publish-test-results",
+  "helpMarkDown": "ms-resource:loc.helpMarkDown",
+  "category": "Test",
+  "visibility": [
+    "Build",
+    "Release"
+  ],
+  "author": "Microsoft Corporation",
+  "version": {
+    "Major": 1,
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV1/task.loc.json]: Make this 0. The task should be in preview mode

File: Tasks/ContainerStructureTestV1/Strings/resources.resjson/en-US/resources.resjson [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+  "loc.friendlyName": "Container Structure Test",
+  "loc.helpMarkDown": "[Learn more about this task](https://go.microsoft.com/fwlink/?LinkID=613742)",
+  "loc.description": "Runs Container structure tests using Google container structure test framework",
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV1/Strings/resources.resjson/en-US/resources.resjson]: Change to "Uses container-structure-test (https://github.com/GoogleContainerTools/container-structure-test) to validate the structure of an image based on four categories of tests - command tests, file existence tests, file content tests and metadata tests"

Comment (navin22) [reply Tasks/ContainerStructureTestV1/Strings/resources.resjson/en-US/resources.resjson]: Done

File: Tasks/ContainerStructureTestV1/task.loc.json (hunk 4 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV1/task.loc.json]: Done

File: Tasks/ContainerStructureTestV1/task.json (hunk 3 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV1/task.json]: I have used it as place holders. I am not using it right now. Will sync up with you and Vikranth on how to use some these.

File: Tasks/ContainerStructureTestV1/task.json (hunk 1 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV1/task.json]: Done

File: Tasks/ContainerStructureTestV1/task.json (hunk 2 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV1/task.json]: Agree, changed it.

File: Tasks/ContainerStructureTestV0/task.json [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            "label": "Image",
+            "required": true,
+            "helpMarkDown": "Image to be tested"
+        },
+        {
+            "name": "configFile",
+            "type": "filePath",
+            "label": "Config Files",
+            "required": true,
+            "helpMarkDown": "Config files path, that contains container structure tests. Either .yaml or .json files"
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV0/task.json]: If we are supporting multiple config files (that is, we are iteratively running the test binary on each of these config files), then mention that glob pattern is supported to allow for specification of multiple config paths

File: Tasks/ContainerStructureTestV0/task.json [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        },
+        {
+            "name": "failTaskOnFailedTests",
+            "type": "boolean",
+            "label": "Fail if there are test failures",
+            "defaultValue": "false",
```
Comment (shashankbarsin) [inline Tasks/ContainerStructureTestV0/task.json]: Is defaultValue: false consistent with how other test tasks run? I don't have much context on the test tasks, that's why I'm asking

File: make-options.json [hunk 8]
```diff
@@ -45,6 +45,7 @@
         "CocoaPodsV0",
         "CondaEnvironmentV0",
         "CondaEnvironmentV1",
+		"ContainerStructureTestV1",
```
Comment (shashankbarsin) [inline make-options.json]: Should this be V0?

File: Tasks/ContainerStructureTestV0/task.json [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                "resizable": "true",
+                "rows": "2"
+            },
+            "visibleRule": "command = build || command = push || command = buildAndPush",
+            "label": "Tags",
+            "groupName": "commands",
+            "helpMarkDown": "A list of tags in separate lines. These tags are used in build, push and buildAndPush commands. Ex:<br><br>beta1.1<br>latest"
+        },
+        {
+            "name": "image",
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/ContainerStructureTestV0/task.json]: Oh okay, I haven't started using those inputs yet. Will change it. 
Thought, we will be having images inside repository.

File: Tasks/ContainerStructureTestV0/task.json (hunk 7 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV0/task.json]: Yes. In other test tasks also we have it as false.

File: Tasks/ContainerStructureTestV0/task.json (hunk 6 above)
Comment (navin22) [reply Tasks/ContainerStructureTestV0/task.json]: Currently I am not supporting it. Will add this functionality in next iteration.

File: Tasks/ContainerStructureTestV0/containerstructuretest.ts [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    let taskResult = true;
+    try {
+        const osType = tl.osType().toLowerCase();
+        tl.debug(`Os Type: ${osType}`);
+        telemetryData["OsType"] = osType;
+
+        const testFilePath = tl.getInput('configFile', true);
+        const repository = tl.getInput('repository', true);
+        let tagsInput = tl.getInput('tag');
+        const tag = tagsInput == null || tagsInput == "" ? "latest" : tagsInput;
```
Comment (vithati) [inline Tasks/ContainerStructureTestV0/containerstructuretest.ts]: tagsInput? tagsInput : "latest"

File: Tasks/ContainerStructureTestV0/task.loc.json [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    },
+    {
+      "name": "configFile",
+      "type": "filePath",
+      "label": "ms-resource:loc.input.label.configFile",
+      "required": true,
+      "helpMarkDown": "ms-resource:loc.input.help.configFile"
+    },
+    {
+      "name": "failTaskOnFailedTests",
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/ContainerStructureTestV0/task.loc.json]: This is to be inconsistent with the other Test tasks.

[the remark this answers is not in the document]
Comment (vithati) [reply Tasks/ContainerStructureTestV0/task.loc.json]: If it aligns with other test tasks, then it's ok.

File: Tasks/ContainerStructureTestV0/task.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        "Build",
+        "Release"
+    ],
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 0,
+        "Minor": 1,
+        "Patch": 0
+    },
+    "demands": [],
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/ContainerStructureTestV0/task.json]: Adding check in the code and throwing the appropriate error.

File: .github/CODEOWNERS [hunk 13]
```diff
@@ -291,6 +291,8 @@ Tasks/PublishTestResultsV1/*    @smalpani
 
 Tasks/PublishTestResultsV2/*    @SadagopanRajaram
 
+Tasks/ContainerStructureTestV0/*    @SadagopanRajaram
```
Comment (vithati) [inline .github/CODEOWNERS]: add your name also.

File: Tasks/ContainerStructureTestV0/containerstructuretest.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        taskResult = false;
+    } finally {
+        tl.setResult(taskResult ? tl.TaskResult.Succeeded : tl.TaskResult.Failed, '');
+        telemetryData["TaskResult"] = taskResult;
+        publishTelemetry();
+        tl.debug("Finished Execution");
+    }
+}
+
+async function dockerLogin(connection: ContainerConnection): Promise<any> {
```
Comment (vithati) [inline Tasks/ContainerStructureTestV0/containerstructuretest.ts]: Remove this. Not used anywhere.

Comment (navin22) [reply Tasks/ContainerStructureTestV0/containerstructuretest.ts]: Yeah, forgot to delete it.

**Pull request merged**